src/remoteClient/clientContextImpl.cpp

```diff
@@ -4556,8 +4608,44 @@ class InternalClientContextImpl :
     Configuration::shared_pointer m_configuration;
 
     TransportRegistry::transportVector_t m_flushTransports;
+
+    friend class BeaconCleanupHandler;
 };
 
+
+BeaconCleanupHandler::BeaconCleanupHandler(InternalClientContextImpl& impl, osiSockAddr addr) :
+    m_callback(std::make_shared<Callback>(*this)),
+    m_from(addr),
+    m_impl(impl)
+{
+    m_impl.m_timer->schedulePeriodic(m_callback, maxBeaconLifetime, maxBeaconLifetime);
+}
+
+BeaconCleanupHandler::~BeaconCleanupHandler()
+{
+    m_impl.m_timer->cancel(m_callback);
+}
+
+void BeaconCleanupHandler::Callback::callback()
+{
+    epicsTimeStamp ts;
+    epicsTimeGetCurrent(&ts);
+    if (epicsTimeDiffInSeconds(&ts, &m_handler.m_lastTime) > maxBeaconLifetime)
+        m_handler.remove();
```

Comment:
I don't see any need for precision here.  And calling `epicsTimeGetCurrent()` potentially 20,000 times could be avoided.

Replace `epicsTimeStamp m_lastTime` with a integer counter.  Set the timer to eg `maxBeaconLifetime/2`.  Increment the counter on expiration.  If it reaches 3, then `remove()`.  Have `touch()` zero the counter.

That said, this is an optimization.  What you have here will work.  So please let me know if you will not have to for this change.

Reply:
I don't have a good feel for what this cap should be.  My most recent solid point of reference is NSLS2 circa 2015, when I snooped on beacon traffic to identify ~1000 CA server instances running on ~100 hosts.  Well short of 20k, but only a factor of 2 from 2048.

PVXS currently tracks up to 20k addresses.  I picked this number from thin air.  Although I opted for a larger number because PVXS uses fewer resources per server ( ~= 256 bytes per).

Reply:
Assuming this cap is the maximum number of servers that could be generating beacons visible to a client, the APS accelerator currently has 1136 IOCs in production (mostly CA currently) spread across 5 VLANs, with 881 IOCs in the largest VLAN. I don't see our total number increasing beyond 2048 IOCs, but we do have 13 VMs that run multiple soft IOCs (the largest VM has 214) so the number of visible servers isn't limited to the subnet size. Presumably configuring PVA servers through a shared multicast address would include the servers from multiple subnets too.

I'm fine with the cap being 2048 for pvAccessCPP. It might be worth having the PVXS cap be configurable by an env-var at some point, how many IOCs is ITER expecting to run?